### PR TITLE
TLS1.3: random session_id for ClientHello

### PR DESCRIPTION
**Pull request by tomato42**
### Description

in TLS 1.3 the middle box compatibility mode
requires the client to send an unpredictable session_id
and the server to echo it

for that, we need to actually set it to some value when
the extensions indicate that it's a TLS 1.3 ClientHello

### Motivation and Context

tests need to verify if the server follows the middlebox compatibility requirements if client asks for it

### Checklist

- [X] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [X] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
  - n/a
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
  - n/a
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [x] NSS
  - [x] GnuTLS
- [x] required version of tlslite-ng updated in requirements.txt and README.md
  - n/a

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/391)

### CODE REVIEW

File: tlsfuzzer/messages.py [hunk 1]
```diff
@@ -413,8 +413,17 @@ def generate(self, state):
...
         if self.random:
             state.client_random = self.random
-        if self.session_id is None:
+        if self.session_id is None and state.session_id:
             self.session_id = state.session_id
+        if self.session_id is None and self.extensions \
+                and ExtensionType.supported_versions in self.extensions:
+            # in TLS 1.3, the server should reply with CCS (middlebox compat
+            # mode) only when client sends a session_id
+            self.session_id = getRandomBytes(16)
```
Comment (ueno) [inline tlsfuzzer/messages.py]: why this is 16-octet long while the spec says "legacy_session_id: ... a client not offering a pre-TLS 1.3 session MUST generate a new 32-byte value"?

File: tlsfuzzer/messages.py [hunk 2]
```diff
@@ -413,8 +413,17 @@ def generate(self, state):
...
-        if self.session_id is None:
+        if self.session_id is None and state.session_id:
             self.session_id = state.session_id
+        if self.session_id is None and self.extensions \
+                and ExtensionType.supported_versions in self.extensions:
+            # in TLS 1.3, the server should reply with CCS (middlebox compat
+            # mode) only when client sends a session_id
+            self.session_id = getRandomBytes(32)
+        # if still unset, set to default value
+        if not self.session_id:
```
Comment (ueno) [inline tlsfuzzer/messages.py]: nit: `is None` might be better for consistency?

Comment (tomato42) [reply tlsfuzzer/messages.py]: could be, but the difference between None and empty string bit me in previous version of the PR, so I prefer the more lenient version here

**Comment by tomato42** [comment]
Thank you!

**Pull request merged**